sensor/kubernetes/crs/crs.go

```diff
@@ -0,0 +1,308 @@
+	fetchSensorDeploymentOwnerRefBackoff = wait.Backoff{
+		Duration: 10 * time.Millisecond,
+		Factor:   3,
+		Jitter:   0.1,
+		Steps:    10,
+		Cap:      6 * time.Minute,
+	}
+)
+
+// EnsureClusterRegistered initiates the CRS based cluster registration flow if
+//
+//  1. no Kubernetes secret named `sensor-tls` exists -- existence of this secret
+//     would indicate that service-specific secrets have already been set up using
+//     e.g. an init-bundle.
+//
+//     and
+//
+//  2. no Kubernetes secret named `tls-sensor` exists -- existence of this secret
+//     would indicate that service-specific secrets have already been retrieved via
+//     the CRS-flow (or a regular follow-up certificate rotation).
+func EnsureClusterRegistered() error {
+	log.Infof("Ensuring Secured Cluster is registered.")
+	clientconn.SetUserAgent(fmt.Sprintf("%s CRS", clientconn.Sensor))
+
+	// Check if legacy sensor service certificate, e.g. created via init-bundle, exists.
+	if legacySensorServiceCert := os.Getenv(crs.LegacySensorServiceCertEnvName); legacySensorServiceCert != "" {
+		log.Infof("Legacy sensor service certificate available, skipping CRS-based cluster registration.")
+		return nil
+	}
+
+	// Check if modern sensor service certificate, e.g. created by CRS-based cluster registration or
+	// refreshed by periodic cert rotation, exists.
+	if sensorServiceCert := os.Getenv(crs.SensorServiceCertEnvName); sensorServiceCert != "" {
+		log.Infof("Sensor service certificate available, skipping CRS-based cluster registration.")
+		return nil
+	}
+
+	log.Infof("No sensor service certificate found, initiating CRS-based cluster registration.")
+	if err := registerCluster(); err != nil {
+		return errors.Wrap(err, "registering secured cluster")
+	}
+	log.Info("CRS-based cluster registration complete.")
+
+	return nil
+}
+
+// registerCluster implements the CRS-based registration flow for secured clusters.
+// It
+//   - retrieves the CRS secret
+//   - unpacks the contained mTLS certificate+key pair for the REGISTRANT_SERVICE identity from the
+//     CRS and stores it in /run/secrets/stackrox.io/certs (which must be a memory-backed
+//     emptyDir) so that they will be picked up automatically for our mTLS authentication towards Central.
+//   - connects to Central and sends a sensorHello message for this new secured cluster
+//   - expects to receive a centralHello response containing newly issued service certificates+keys
+//     which are then stored as Kubernetes secrets named `tls-cert-<service slug name>`.
+func registerCluster() error {
+	ctx := context.Background()
+
+	log.Infof("Trying to load Cluster Registration Secret.")
+	crs, err := crs.Load()
+	if err != nil {
+		log.Errorf("failed to load CRS: %v", err)
+		return errors.Wrap(err, "loading CRS")
+	}
+
+	err = temporarilyStoreRegistrantSecret(crs)
+	if err != nil {
+		return errors.Wrap(err, "preparing registration secret for mTLS authentication")
+	}
+
+	// Connect to Central.
```

Comment:
I think this and the following two comments don't contribute anything, the function calls are self-explanatory.

Reply:
done